From 57e36a88e7dd9ac86f80de25dd6e22848fc7cd9c Mon Sep 17 00:00:00 2001
From: "Darrick J. Wong" <djwong@kernel.org>
Date: Mon, 13 Jul 2026 23:03:44 -0700
Subject: [PATCH] xfs: resample the data fork mapping after cycling ILOCK

commit 2f4acd0fcd862e22eab45690ec2c08c80b6ef2e7 upstream.

xfs_reflink_fill_{cow_hole,delalloc} are both presented with an inode,
a data fork mapping, and a cow fork mapping.  Unfortunately, these two
helpers cycle the ILOCK to grab a transaction, which means that the
mappings are stale as soon as we reacquire the ILOCK.  Currently we
refresh the cow fork mapping by re-calling xfs_find_trim_cow_extent, but
we don't refresh the data fork mapping beforehand, which means that the
xfs_bmap_trim_cow in that function queries the refcount btree about the
wrong physical blocks and returns an inaccurate value in *shared.

If *shared is now false, the directio write proceeds with a stale data
fork mapping.  Fix this by querying the data fork mapping if the
sequence counter changes across the ILOCK cycle.

Cc: hch@lst.de
Cc: stable@vger.kernel.org # v4.11
Fixes: 3c68d44a2b49a0 ("xfs: allocate direct I/O COW blocks in iomap_begin")
Signed-off-by: "Darrick J. Wong" <djwong@kernel.org>
Reviewed-by: Christoph Hellwig <hch@lst.de>
Reviewed-by: Carlos Maiolino <cmaiolino@redhat.com>
Signed-off-by: Carlos Maiolino <cem@kernel.org>
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
(cherry picked from commit e705d81a7193dd19e69b8e2bad4696d78a4ea075)

Backports to v4.19, conflicts due to missing:

- f273387b0485 ("xfs: refactor reflink functions to use
xfs_trans_alloc_inode"): slight context differences with use of
xfs_trans_alloc_inode instead of open-coding the locking

- d62113303d69 ("xfs: Fix false ENOSPC when performing direct write on a
delalloc extent in cow fork"), function split into two helpers to prevent
enospace conditions with delalloc -> we only need to add one hunk in the
non-split xfs_reflink_allocate_cow function.

Signed-off-by: Quentin Casasnovas <quentin.casasnovas@vates.tech>
---
 fs/xfs/xfs_reflink.c | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/fs/xfs/xfs_reflink.c b/fs/xfs/xfs_reflink.c
index 42ea7bab9144..e3160ce7a8c8 100644
--- a/fs/xfs/xfs_reflink.c
+++ b/fs/xfs/xfs_reflink.c
@@ -405,6 +405,7 @@ xfs_reflink_allocate_cow(
 	xfs_fileoff_t		offset_fsb = imap->br_startoff;
 	xfs_filblks_t		count_fsb = imap->br_blockcount;
 	struct xfs_trans	*tp;
+	unsigned int            seq_before = READ_ONCE(ip->i_df.if_seq);
 	int			nimaps, error = 0;
 	bool			found;
 	xfs_filblks_t		resaligned;
@@ -435,6 +436,23 @@ xfs_reflink_allocate_cow(
 	if (error)
 		goto out_trans_cancel;
 
+	/*
+	 * The data fork mapping may have changed while we dropped the ILOCK
+	 * (a racing O_DIRECT writer under IOLOCK_SHARED can complete a full
+	 * CoW cycle including xfs_reflink_end_cow(), which remaps this offset
+	 * and drops the refcount of the old shared block).  Re-read it so the
+	 * shared-status recheck below and the caller's in-place iomap both
+	 * operate on the current mapping rather than a stale physical block.
+	 */
+	if (seq_before != READ_ONCE(ip->i_df.if_seq)) {
+		nimaps = 1;
+		error = xfs_bmapi_read(ip, imap->br_startoff,
+				       imap->br_blockcount, imap, &nimaps, 0);
+		if (error)
+			goto out_trans_cancel;
+	}
+
+
 	/*
 	 * Check for an overlapping extent again now that we dropped the ilock.
 	 */
